train.py

```diff
@@ -0,0 +1,99 @@
+import os
+import sys
+import time
+import argparse
+
+import torch
+import psutil
+import imageio
+import numpy as np
+from torchvision.utils import save_image
+
+from utils import get_interpolations
+from models.autoencoder import Autoencoder
+from train_utils import test_epoch, train_epoch
+
+parser = argparse.ArgumentParser(description="Main function to call training for different AutoEncoders")
+parser.add_argument("--batch-size", type=int, default=128, metavar="N", help="input batch size for training (default: 128)")
+parser.add_argument("--epochs", type=int, default=10, metavar="N", help="number of epochs to train (default: 10)")
+parser.add_argument("--no-cuda", action="store_true", default=False, help="enables CUDA training")
+parser.add_argument("--seed", type=int, default=42, metavar="S", help="random seed (default: 1)")
+parser.add_argument(
+    "--log-interval", type=int, default=10, metavar="N", help="how many batches to wait before logging training status"
+)
+parser.add_argument(
+    "--embedding-size", type=int, default=32, metavar="N", help="how many batches to wait before logging training status"
+)
+parser.add_argument("--results_path", type=str, default="results/", metavar="N", help="Where to store images")
+parser.add_argument("--model", type=str, default="AE", metavar="N", help="Which architecture to use")
+parser.add_argument("--dataset", type=str, default="MNIST", metavar="N", help="Which dataset to use")
+
+args = parser.parse_args()
+args.cuda = not args.no_cuda and torch.cuda.is_available()
+torch.manual_seed(args.seed)
+
+ae = Autoencoder(args)
+architectures = {"AE": ae}
+
+print(args.model)
+
+if __name__ == "__main__":
+    try:
+        os.stat(args.results_path)
+    except FileNotFoundError:
+        os.mkdir(args.results_path)
+
+    try:
+        autoenc = architectures[args.model]
+    except KeyError:
+        print("---------------------------------------------------------")
+        print("Model architecture not supported. ", end="")
+        print("---------------------------------------------------------")
+        sys.exit()
+
+    try:
+        total_start_time = time.time()
+        process = psutil.Process()
+        max_memory_usage = 0
+        with open(os.path.join(args.results_path, f"training_time_{args.model}_{args.dataset}.txt"), "w") as f:
+            optimizer = torch.optim.Adam(autoenc.model.parameters(), lr=1e-3)
+            for epoch in range(1, args.epochs + 1):
```

Comment:
We should strive to have the indentation level as low as possible. Here you should be able to get rid of `try:` and `with open() as f:`.

Please check out https://wandb.ai/ and set it up to track any metrics - it's much more convenient than writing text files with custom names. If you had to store it as a file, you can still avoid the nested indent:

```python
performance_stats = []
for epoch in range(epochs):
  ...
  performance_stats["max_memory_usage"] = 7

with open("performance_stats") as f:
  f.write(performance_stats)
```
  
For the `try` block I'd prefer something like:

```python
try:
  train_model()
except KeyboardInterrupt:
  print("hi")
```

This would also force you to have the training logic well separated.

Also, catching keyboard interrupt is a nice touch, but in my experience when I hit ctr+c on a run I don't really want any code to be executed after that :)